core/src/main/java/io/grpc/internal/ServerCallImpl.java

```
@@ -338,6 +345,9 @@ private void messagesAvailableInternal(final MessageProducer producer) {
           }
           message.close();
         }
+      } catch (TooLongDecompressedMessageException e) {
+        this.call.exceptionStatus = e.getStatus();
+        this.call.handleInternalError(e);
```

The introduction of this.call.exceptionStatus is an unnecessary complication. It introduces one more field that handleInternalError checks for and ignores the exception argument in that case. 
pachenko's suggestion of TooLongDecompressedMessageException was for having a separate catch in ServerImpl.JumpToApplicationThreadServerStreamListener#messagesAvailable, which we are not doing now, so it doesn't apply. 
It could just have been a StatusRuntimeException with RESOURCE_EXHAUSTED and checking for the message description to distinguish too large message parsing error from other types of RESOURCE_EXHAUSTED like you suggested before, and leave `handleInternalError` unchanged.

There was also the suggestion to do `call.canceled = true;` in the exception handling for parse error due to too large size, that is missing.